src/optimizer/util.ts

```diff
@@ -0,0 +1,303 @@
+): AstExpression {
+    const result = createAstNode({
+        kind: "op_unary",
+        op: op,
+        operand: operand,
+        loc: DUMMY_LOCATION,
+    });
+    return result as AstExpression;
+}
+
+export function makeBinaryExpression(
+    op: AstBinaryOperation,
+    left: AstExpression,
+    right: AstExpression,
+): AstExpression {
+    const result = createAstNode({
+        kind: "op_binary",
+        op: op,
+        left: left,
+        right: right,
+        loc: DUMMY_LOCATION,
+    });
+    return result as AstExpression;
+}
+
+// Checks if the top level node is a binary op node
+export function checkIsBinaryOpNode(ast: AstExpression): boolean {
+    return ast.kind === "op_binary";
+}
+
+// Checks if top level node is a binary op node
+// with a non-value node on the left and
+// value node on the right
+export function checkIsBinaryOp_NonValue_Value(ast: AstExpression): boolean {
+    if (ast.kind === "op_binary") {
+        return !isValue(ast.left) && isValue(ast.right);
+    } else {
+        return false;
+    }
+}
+
+// Checks if top level node is a binary op node
+// with a value node on the left and
+// non-value node on the right
+export function checkIsBinaryOp_Value_NonValue(ast: AstExpression): boolean {
+    if (ast.kind === "op_binary") {
+        return isValue(ast.left) && !isValue(ast.right);
+    } else {
+        return false;
+    }
+}
+
+// bigint arithmetic
+
+// precondition: the divisor is not zero
+// rounds the division result towards negative infinity
+export function divFloor(a: bigint, b: bigint): bigint {
+    const almostSameSign = a > 0n === b > 0n;
+    if (almostSameSign) {
+        return a / b;
+    }
+    return a / b + (a % b === 0n ? 0n : -1n);
+}
+
+export function abs(a: bigint): bigint {
+    return a < 0n ? -a : a;
+}
+
+export function sign(a: bigint): bigint {
+    if (a === 0n) return 0n;
+    else return a < 0n ? -1n : 1n;
+}
+
+// precondition: the divisor is not zero
+// rounds the result towards negative infinity
+// Uses the fact that a / b * b + a % b == a, for all b != 0.
+export function modFloor(a: bigint, b: bigint): bigint {
+    return a - divFloor(a, b) * b;
+}
+
+// Test equality of ASTExpressions.
+export function areEqualExpressions(
+    ast1: AstExpression,
+    ast2: AstExpression,
+): boolean {
+    switch (ast1.kind) {
+        case "null":
+            return ast2.kind === "null";
+        case "boolean":
+            return ast2.kind === "boolean" ? ast1.value === ast2.value : false;
+        case "number":
+            return ast2.kind === "number" ? ast1.value === ast2.value : false;
+        case "string":
+            return ast2.kind === "string" ? ast1.value === ast2.value : false;
+        case "id":
+            return ast2.kind === "id" ? ast1.text === ast2.text : false;
+        case "method_call":
+            if (ast2.kind === "method_call") {
+                return (
+                    ast1.method.text === ast2.method.text &&
```

Comment:
Let's use the identifier comparison function `eqNames` here and elsewhere:
```suggestion
                    eqNames(ast1.method, ast2.method) &&
```